### PR TITLE
docs: add concept document on vaults

### PR DESCRIPTION
**Pull request by mrinalwadhwa**

### CODE REVIEW

File: documentation/concepts/vaults.md [hunk 1]
```diff
@@ -1,3 +1,30 @@
...
+an abstract `Vault` interface.
+
+A concrete implementation of the `Vault` interface is called an Ockam
+Vault. Over time there will be many such implementations.
+
+![Vaults](./assets/vaults/vaults.svg)
+
+The `Vault` interface exposes functions for generating and using
+cryptographic keys, secure random numbers, key derivation, hashing,
+massage authentication, digital signatures, Diffie-Hellman key exchange,
```
Comment (mtm0183) [inline documentation/concepts/vaults.md]: message not massage

File: documentation/concepts/vaults.md [hunk 2]
```diff
@@ -1,3 +1,30 @@
...
+authenticated encryption etc.
+
+Our general design principle here is to minimize cryptographic agility
+and is aim for
+[well oiled joints](https://www.imperialviolet.org/2016/05/16/agility.html).
```
Comment (mattgreg) [inline documentation/concepts/vaults.md]: My suggestion is to add a sentence that gives an overview of well oiled joints for now, and expand upon this concept later, in the context of our specific application of it.  In the mean time I feel that the link is a distraction.

Comment (mrinalwadhwa) [reply documentation/concepts/vaults.md]: noted.

File: documentation/concepts/vaults.md [hunk 3]
```diff
@@ -1,3 +1,30 @@
 ```yaml
 title: Vaults
 ```
+
+Various protocols, implemented in Ockam, depend on a variety of standard
```
Comment (mattgreg) [inline documentation/concepts/vaults.md]: Confusing sentence construction. 

"Ockam's protocols depend upon....

File: documentation/concepts/vaults.md [hunk 4]
```diff
@@ -1,3 +1,30 @@
...
+cryptographic primitives or building blocks. Depending on the environment,
+these building blocks may be provided by a software implementation or a
+cryptographically capable hardware component.
+
+In order to support a variety of cryptographically capable hardware we
+wish to maintain loose coupling between a protocol and how a specific
```
Comment (mattgreg) [inline documentation/concepts/vaults.md]: strike 'wish to'

**Pull request merged**